### PR TITLE
Control Parallel Git Invocation

### PR DESCRIPTION
**Pull request by scbedd**
This PR ensures that we will ever run git commands in **serial** against a given working directory. Controlling per directory means that concurrent tests from various service areas will still be able to work alongside each other.

@benbp what do you think about this? I like that it is _entirely_ atomic to the class. When we move this guy to a common core, this functionality should definitely merge IMO.

### CODE REVIEW

**Comment by benbp** [comment]
@scbedd @hallipr also has some task queue logic built up, but it's a little more dependent on external services which we don't want here (though the task queue message provider could be local). Looks good though.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitProcessHandler.cs [hunk 1]
```diff
@@ -106,43 +110,52 @@ public virtual CommandResult Run(string arguments, string workingDirectory)
...
-                DebugLogger.LogInformation($"git {arguments}");
-                var process = Process.Start(processStartInfo);
-                string stdOut = process.StandardOutput.ReadToEnd();
-                string stdErr = process.StandardError.ReadToEnd();
-                process.WaitForExit();
+                value = new TaskQueue();
+                // only add a new task queue if it doesn't exist
+                // if it does exist, just use the old one so that pending tasks will still be able to complete
+                AssetTasks.AddOrUpdate(workingDirectory, value, (key, oldValue) => oldValue);
+            }
```
Comment (hallipr) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitProcessHandler.cs:119]: There's still a race condition here.  If AssetTasks.TryGetValue returns false, but AddOrUpdate hits the Update func, you'll keep the old value in the dictionary but continue to use the new value created on line 115 for the enqueue statement on  line 121

Comment (hallipr) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitProcessHandler.cs:119]: You could rewrite using GetOrAdd like:

```cs
            // only add a new task queue if it doesn't exist
            var value = AssetTasks.GetOrAdd(workingDirectory, () => new TaskQueue());
```

Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitProcessHandler.cs:119]: Thanks for the feedback. Definitely added.

**Review by JimSuplizio — APPROVED** [review]
This is goodness. @hallipr already noticed the potential issue that you've already fixed so I think we're good here.

**Comment by scbedd** [comment]
Added in further feedback from [and the repo tests are passing](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1848042&view=results) using the new methodology!

**Pull request merged**